# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def addTwo(x):
    return x*2
print(addTwo(4))

#Lambda version...

print((lambda x: x+2)(4))

#Using a variable...

a_num=4
print((lambda x: x+2)(a_num))

8
6
6


#### Saving to a Variable

In [4]:
f_test= lambda x: x+2

f_test(4)

6

#### Multiple Inputs

In [9]:
print((lambda x,y,z: x*y*z)(3,5,8))

x_test=(lambda x,y,z: x*y*z)
x_test(3,5,8)

120


120

#### Passing a Lambda into a Function

In [7]:
def multiply(f, num):
    '''
    f expect a lambda function
    num expect a number
    '''
    
    return f(num)

multiply(lambda x: x* x,4)

16

#### Returning a Lambda from a Function

In [11]:
def returnLamb(b,c):
    return lambda x,a: x+a+b+c
r_lamb=returnLamb(4,3)
print(r_lamb(2,3))

12


#### If Statements within Lambdas

In [13]:
  f_condition= lambda num: num*2 if num>10 else num+2

print(f_condition(3))
print(f_condition(12))

5
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [14]:
f= lambda num: num**2
print(f(2))

4


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [39]:
#map(function, iterable)
def squared(num, num2):
        if num<10 and num2<10:
            return num**2, num2**2
        else:
            return num, num2
numbers=[2,4,5,6]
numbers1=[5,8,2,6]
squared_nums_map=list((map(squared,numbers, numbers1)))
print(squared_nums_map)

[(4, 25), (16, 64), (25, 4), (36, 36)]


#### Using Lambda's with Map

In [40]:
squared_nums_lambda=list(map(lambda x,y: (x**2, y**2)if x<10 and y<10 else(x,y), numbers, numbers1))
print(squared_nums_lambda)
print(list(map(lambda x,y: (x**2, y**2)if x<10 and y<10 else(x,y), numbers, numbers1)))


[(4, 25), (16, 64), (25, 4), (36, 36)]
[(4, 25), (16, 64), (25, 4), (36, 36)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [41]:
nums=list(map(lambda x, y: (x*2-1,y*2-1), numbers, numbers1))
print(nums)

[(3, 9), (7, 15), (9, 3), (11, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [35]:
names=["Sam","Ali","Ahmed","Fay"]

def a_names(name):
    if name[0].lower()=='a':
        return True
    else:
        return False
new_names=list(filter(a_names, names))
print(new_names)

['Ali', 'Ahmed']


#### Using Lambda's with Filter()

In [37]:
new_names=list(filter(lambda name: True if name[0].lower()=='a' else False, names))
print(new_names)

['Ali', 'Ahmed']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [42]:
import statistics

mean_num=(statistics.mean(numbers1))

nums_below_mean=list(filter(lambda num: True if num < mean_num else False,numbers1))

print(nums_below_mean)



[5, 2]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [45]:
from functools import reduce
list_1=[1,2,3,4,5,6]

def addNums(num, num1):
    return num+num1
result_add=reduce(addNums, list_1)
print(result_add)


def subNums(num, num1):
    return num-num1
result_sub=reduce(subNums, list_1)
print(result_sub)

21
-19


#### Using Lambda's with Reduce()

In [47]:
print(reduce(lambda num, num2: num+num2, list_1))

21


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [52]:
 my_list=[1,2,3,4]
multi=reduce(lambda num,num1: num*num1, my_list)
print(multi)


24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [73]:
def addNums(num):
    #Here is our base case...
    if num <=1:
        print("addnums(1)= 1")
        return num
    else:
        print(f"addNums({num})={num}+addNums{num-1})")
        return num + addNums(num-1)
        
addNums(7)

addNums(7)=7+addNums6)
addNums(6)=6+addNums5)
addNums(5)=5+addNums4)
addNums(4)=4+addNums3)
addNums(3)=3+addNums2)
addNums(2)=2+addNums1)
addnums(1)= 1


28

#### Writing a Factorial Function

In [72]:
def factorial(num):
    if num<=1:
        return 1
    else:
        return num*factorial(num-1)
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [63]:
def subNums(num):
    if num<=1:
        return 1
    else:
        return num-subNums(num-1)
subNums(8)
    

4

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [70]:
def squareNums(nums):
    for i in nums:
        yield(i*i)
my_nums=squareNums([1,2,3,4,5,6])
for num in my_nums:
    print(num)
    
my_nums= (x*x for x in [1,2,3,4,5])
print (list(my_nums))


1
4
9
16
25
36
[1, 4, 9, 16, 25]


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def squareNums(nums):
    for i in nums:
        yield(i*i)
my_nums=squareNums([1,2,3,4,5,6])
for num in my_nums:
    print(num)
    

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
empty_strings=[" ","","  "]
for i in empty_strings:
    while (i) in places:
            places.remove(i)
print(places)

   




### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



author.sort(key=lambda a: a.split()[1])

print(author)






### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
newplaces = list(map(lambda data: (data[0],  (9/5) * data[1]+ 32), places))
print(newplaces)




### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [ ]:
def F(i):
    if i==0:
        return 0
    elif i==1:
        return 1
    else:
        return F(i-2)+F(i-1)
for x in range (5):
      print (F(x))
